README.md
# JavaScript Base-N Id Generator _(bnid)_

[![Node.js CI](https://github.com/digitalbazaar/bnid/workflows/Node.js%20CI/badge.svg)](https://github.com/digitalbazaar/bnid/actions?query=workflow%3A%22Node.js+CI%22)

> A JavaScript library for Web browsers and Node.js apps to generate random
> ids and encode and decode them using various base-N encodings.

## Table of Contents

- [Background](#background)
- [Install](#install)
- [Usage](#usage)
- [API](#api)
- [CLI](#cli)
- [Contribute](#contribute)
- [Commercial Support](#commercial-support)
- [License](#license)

## Background

This library provides tools for Web and Node.js to generate random ids and
encode and decode them in various base-N encodings.

## Install

### NPM

```
npm install bnid
```

### Git

To install locally (for development):

```
git clone https://github.com/digitalbazaar/bnid.git
cd bnid
npm install
```

## Usage

The interface follows the [TextEncoder][]/[TextDecoder][] interfaces and
provides [IdEncoder](#idencoder) and [IdDecoder](#iddecoder) classes. Instances
can be configured with reusable encodings and other parameters. The encoder
operates on a `Uint8Array` of input bytes. A [IdGenerator](#idgenerator) class
is provided that can be used to generate random ids with selectable bit
lengths, but any id data can be used.

The encoder and decoder support various encodings and options:

- `base16`/`hex`, `base16upper`: The simple [Base16][].
- `base58`/`base58btc`: The [Base58][] Bitcoin alphabet as supported by
  [base58-universal][].
- Optional [multibase][] type prefix.
- Fixed bit length. This is useful to ensure the output id length is constant
  even when the id starts with an arbitrary number of zeros.

The fixed length options can be important when using encodings that have
variable length outputs depending on the input length. `base58btc` is and
example. When encoding, the `fixedLength` or `fixedBitLength` options can be
used to force the output to be a constant length. When decoding, the
`fixedBitLength` options can be used to ensure a constant length array of
bytes.

### Generate an ID

To generate a default [Base58][], 128 bit, non-fixed-length, multibase encoded
id:

```js
import {generateId} from 'bnid';

const id = await generateId();
```

To generate a [Base58][], 128 bit, fixed-length id:

```js
import {generateId} from 'bnid';

const id = await generateId({
  fixedLength: true
});
```

### Reusable Components

Some setup overhead can be avoided by using the component `IdGenerator` and
`IdEncoder` classes.

```js
import {IdGenerator, IdEncoder} from 'bnid';

// 64 bit random id generator
const generator = new IdGenerator({
  bitLength: 64
});
// base58, multibase, fixed-length encoder
const encoder = new IdEncoder({
  encoding: 'base58',
  fixedLength: true,
  multibase: true
});
const id1 = encoder.encode(await generator.generate());
const id2 = encoder.encode(await generator.generate());
```

### Reusable Components

Some setup overhead can be avoided by using the component `IdGenerator` and
`IdEncoder` classes.

```js
import {IdGenerator, IdEncoder, IdDecoder} from 'bnid';

// 64 bit random id generator
const generator = new IdGenerator({
  bitLength: 64
});
// base58, multibase, fixed-length encoder
const encoder = new IdEncoder({
  encoding: 'base58',
  fixedLength: true,
  multibase: true
});
const id1 = encoder.encode(await generator.generate());
// => "z..."
const id2 = encoder.encode(await generator.generate());
// => "z..."

const decoder = new IdDecoder({
  fixedBitLength: 64,
  multibase: true
});
const id1bytes = decoder.decode(id1);
// => Uint8Array([...])
const id2bytes = decoder.decode(id2);
// => Uint8Array([...])
```

## API

### `generateId(options)`

Generate a string id. See `IdGenerator` and `IdEncoder` for options.

### `decodeId(options)`

Decode the options.id string. See `IdDecoder` for other options.

### `IdGenerator`

An `IdGenerator` generates an array of id bytes.

#### `constuctor(options)` / `constructor(bitLength)`

Options:
- `bitLength`: Number of id bits. Must be multiple of 8. (default: 128)

#### `generate()`

Generate random id bytes.

### `IdEncoder`

An `IdEncoder` encodes an array of id bytes into a specific encoding.

#### `constuctor(options)`

Options:
- `encoding`: Output encoding. (default: `base58`)
  - `base16`/`base16upper`/`hex`: base16 encoded string.
  - `base58`/`base58btc`: base58btc encoded string.
- `fixedLength`: `true` to ensure fixed output length. (default: false)
- `fixedBitLength`: fixed output bit length or 0 to base on input byte size.
  (default: 0)
- `multibase`: `true` to use multibase encoding. (default: `true`)
- `multihash`: `true` to use multihash encoding. (default: `false`)

#### `encode(bytes)`

Encode id bytes into a string.

### `IdDecoder`

An `IdDecoder` decodes a specific encoding into an array of bytes representing
an ID.

#### `constuctor(options)`

Options:
- `encoding`: Input encoding. Ignored if `multibase` is `true`. (default:
  `base58`)
  - Same options as for `IdEncoder`.
- `fixedBitLength`: fixed output bit length.  (default: none)
- `multibase`: `true` to use multibase encoding to detect id format. (default:
  `true`)
- `multihash`: `true` to use multihash encoding. (default: `false`)
- `expectedSize`: Expected size for multihash-encoded ID bytes. Use `0` to
  disable size check. (default: 32)

#### `decode(id)`

Decode id string into bytes.

### `minEncodedIdBytes(options)`

Minimum number of bytes needed to encode an id of a given bit length.

Options:
- `encoding`: Encoding. (default: `base58`)
- `bitLength`: Number of id bits. (default: 128)
- `multibase`: Account for multibase encoding. (default: true)

### `maxEncodedIdBytes(options)`

Maximum number of bytes needed to encode an id of a given bit length.

Options:
- `encoding`: Encoding. (default: `base58`)
- `bitLength`: Number of id bits. (default: 128)
- `multibase`: Account for multibase encoding. (default: true)

### `generateSecretKeySeed(options)`

`generateSecretKeySeed()` and `decodeSecretKeySeed()` methods are for creating
and decoding secret key pair seeds which can be used to generate public key
based identifiers.

`generateSecretKeySeed()` generates a secret key seed encoded as a string that
can be stored and later used to generate a key pair. The public key from
the key pair can be used as an identifier. The encoded key seed MUST be kept
secret.

```js
import {generateSecretKeySeed} from 'bnid';
const secretKeySeed = await generateSecretKeySeed();
// Example secretKeySeed: z1Aaj5A4UCsdMpXwdYAReXa4bxWYiKJtdAvB1zMzCHtCbtD
```
Options:
- `encoding`: Encoding. (default: `base58`)
- `bitLength`: Number of id bits. (default: 32 * 8)
- `multibase`: Account for multibase encoding. (default: true)
- `multihash`: Account for multihash encoding. (default: true)

### `decodeSecretKeySeed(options)`

Decodes an encoded secret key seed into an array of secret key seed bytes
(default size: 32 bytes). Both the encoded key seed and the decoded bytes MUST
be kept secret.

```js
import {decodeSecretKeySeed} from 'bnid';
const secretKeySeed = 'z1Aaj5A4UCsdMpXwdYAReXa4bxWYiKJtdAvB1zMzCHtCbtD';
decoded = decodeSecretKeySeed({secretKeySeed});
// Example decoded:
// Uint8Array(32) [
//    80, 174,  15, 131, 124,  59,   9,  51,
//   145, 129,  92, 157, 157, 172, 161,  79,
//    74,  61, 152, 152,  48, 151,  20,  89,
//   225, 169,  71,  34,  49,  61,  21, 215
// ]
```
Options:
- `secretKeySeed`: The secret key seed to be decoded.
- `multibase`: Account for multibase encoding. (default: true)
- `multihash`: Account for multihash encoding. (default: true)
- `expectedSize`: Expected size for multihash-encoded ID bytes. Use `0` to
  disable size check. (default: 32)

## CLI

A command line interface tool to generate and encode ids is provided in
[`bnid-cli`](https://github.com/digitalbazaar/bnid-cli).

## Contribute

Please follow the existing code style.

PRs accepted.

If editing the README, please conform to the
[standard-readme](https://github.com/RichardLitt/standard-readme) specification.

## Commercial Support

Commercial support for this library is available upon request from
Digital Bazaar: support@digitalbazaar.com

## License

[BSD-3-Clause](LICENSE.md) © Digital Bazaar

[Base16]: https://en.wikipedia.org/wiki/Base16
[Base58]: https://en.wikipedia.org/wiki/Base58
[TextDecoder]: https://developer.mozilla.org/en-US/docs/Web/API/TextEncoder
[TextEncoder]: https://developer.mozilla.org/en-US/docs/Web/API/TextDecoder
[base58-universal]: https://github.com/digitalbazaar/base58-universal
[multibase]: https://github.com/multiformats/multibase


CHANGELOG.md
# bnid ChangeLog

## 3.0.0 - 2022-05-13

### Removed
- **BREAKING**: Move command line client into
  [`bnid-cli`](https://github.com/digitalbazaar/bnid-cli). The primary reason
  is to reduce `bnid` dependencies.

### Changed
- **BREAKING**: Convert to module (ESM).
- **BREAKING**: Require Node.js >=14.
- **BREAKING**: Browser support requires `crypto.getRandomValues()`.
- Update dependencies.
- Lint module.

## 2.1.0 - 2021-11-18

### Added
- Add `generateSecretKeySeed()` and `decodeSecretKeySeed()`.
- Add `multihash` boolean parameter to IdEncoder and IdDecoder. A
  multihash-encoded identifier will use the `identity` multicodec tag (`0x00`).
  Using multihash has the advantage of including the size of the identifier
  which can then be verified to have not changed when decoding.
- Add `expectedSize` optional parameter to IdDecoder for multihash-encoded ID
  bytes. (default: 32).

## 2.0.0 - 2020-05-22

### Changed
- **BREAKING**: `multibase` now defaults to `true` for all APIs.
- Remove travis-ci testing in favor of GitHub Actions.

### Added
- Command line tool `bnid`.
- `min/maxEncodedIdBytes` utiltiies. Useful for variable length encodings such
  as `base58`.

## 1.0.0 - 2020-05-07

### Added
- Add core files.

- See git history for changes previous to this release.


tests/benchmark.js
/*!
* Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
*/
import Benchmark from 'benchmark';

const suite = new Benchmark.Suite();

import {
  IdGenerator, IdEncoder, IdDecoder, generateId, decodeId
} from '../lib/index.js';
import crypto from 'node:crypto';

// shared state
const generator = new IdGenerator();
const encoder = new IdEncoder();
const decoder = new IdDecoder();

const idBytes1 = new Uint8Array(128 / 8);
crypto.randomFillSync(idBytes1);
const id1 = encoder.encode(idBytes1);

suite
  .add('generateId base16 128b not-fixed', {
    defer: true,
    fn: async deferred => {
      await generateId({
        encoding: 'base16',
        bitLength: 128,
        fixedLength: false
      });
      deferred.resolve();
    }
  })
  .add('generateId base58 128b not-fixed', {
    defer: true,
    fn: async deferred => {
      await generateId({
        encoding: 'base16',
        bitLength: 128,
        fixedLength: false
      });
      deferred.resolve();
    }
  })
  .add('generateId base58 128b fixed', {
    defer: true,
    fn: async deferred => {
      await generateId({
        encoding: 'base16',
        bitLength: 128,
        fixedLength: true
      });
      deferred.resolve();
    }
  })
  .add('IdGenerator 128b', {
    defer: true,
    fn: async deferred => {
      await new IdGenerator(128).generate();
      deferred.resolve();
    }
  })
  .add('IdGenerator shared 128b', {
    defer: true,
    fn: async deferred => {
      await generator.generate();
      deferred.resolve();
    }
  })
  .add('IdEncoder shared, static data', {
    fn: () => {
      encoder.encode(idBytes1);
    }
  })
  .add('IdDecoder shared, static data', {
    fn: () => {
      decoder.decode(id1);
    }
  })
  .add('decodeId', {
    fn: () => {
      decodeId(id1);
    }
  })
  .add('encode shared', {
    defer: true,
    fn: async deferred => {
      encoder.encode(await generator.generate());
      deferred.resolve();
    }
  })
  .on('cycle', event => {
    console.log(String(event.target));
  })
  .on('complete', function() {
    console.log('Fastest is ' + this.filter('fastest').map('name'));
  })
  .run();


main.js
/*!
 * Copyright (c) 2020 Digital Bazaar, Inc. All rights reserved.
 */
import {
  encode as base58encoder,
  decode as base58decoder
} from 'base58-universal';
import {
  getRandomBytes,
  bytesToHex,
  bytesFromHex
} from './util.js';

// multihash identity function code
const MULTIHASH_IDENTITY_FUNCTION_CODE = 0x00;

function _calcOptionsBitLength({
  defaultLength,
  // TODO: allow any bit length
  minLength = 8,
  // TODO: support maxLength
  //maxLength = Infinity,
  bitLength
}) {
  if(bitLength === undefined) {
    return defaultLength;
  }
  // TODO: allow any bit length
  if(bitLength % 8 !== 0) {
    throw new Error('Bit length must be a multiple of 8.');
  }
  if(bitLength < minLength) {
    throw new Error(`Minimum bit length is ${minLength}.`);
  }
  // TODO: support maxLength
  //if(bitLength > maxLength) {
  //  throw new Error(`Maximum bit length is ${maxLength}.`);
  //}
  return bitLength;
}

function _calcDataBitLength({
  bitLength,
  maxLength
}) {
  if(maxLength === 0) {
    return bitLength;
  }
  if(bitLength > maxLength) {
    throw new Error(`Input length greater than ${maxLength} bits.`);
  }
  return maxLength;
}

function _bytesWithBitLength({
  bytes,
  bitLength
}) {
  const length = bytes.length * 8;
  if(length === bitLength) {
    return bytes;
  }
  if(length < bitLength) {
    // pad start
    const data = new Uint8Array(bitLength / 8);
    data.set(bytes, data.length - bytes.length);
    return data;
  }
  // trim start, ensure trimmed data is zero
  const start = (length - bitLength) / 8;
  if(bytes.subarray(0, start).some(d => d !== 0)) {
    throw new Error(
      `Data length greater than ${bitLength} bits.`);
  }
  return bytes.subarray(start);
}

const _log2_16 = 4;
function _base16Encoder({bytes, idEncoder}) {
  let encoded = bytesToHex(bytes);
  if(idEncoder.encoding === 'base16upper') {
    encoded = encoded.toUpperCase();
  }
  if(idEncoder.fixedLength) {
    const fixedBitLength = _calcDataBitLength({
      bitLength: bytes.length * 8,
      maxLength: idEncoder.fixedBitLength
    });
    const wantLength = Math.ceil(fixedBitLength / _log2_16);
    // pad start with 0s
    return encoded.padStart(wantLength, '0');
  }
  return encoded;
}

const _log2_58 = Math.log2(58);
function _base58Encoder({bytes, idEncoder}) {
  const encoded = base58encoder(bytes);
  if(idEncoder.fixedLength) {
    const fixedBitLength = _calcDataBitLength({
      bitLength: bytes.length * 8,
      maxLength: idEncoder.fixedBitLength
    });
    const wantLength = Math.ceil(fixedBitLength / _log2_58);
    // pad start with 0s (encoded as '1's)
    return encoded.padStart(wantLength, '1');
  }
  return encoded;
}

export class IdGenerator {
  /**
   * Creates a new IdGenerator instance.
   *
   * An IdGenerator generates an array of id bytes.
   *
   * @param {object} [options] - The options to use.
   * @param {number} [options.bitLength=128] - Number of bits to generate.
   *
   * @returns {IdGenerator} - New IdGenerator.
   */
  constructor({
    bitLength
  } = {}) {
    this.bitLength = _calcOptionsBitLength({
      // default to 128 bits / 16 bytes
      defaultLength: 128,
      // TODO: allow any bit length
      minLength: 8,
      bitLength,
    });
  }

  /**
   * Generate random id bytes.
   *
   * @returns {Uint8Array} - Array of random id bytes.
   */
  async generate() {
    const buf = new Uint8Array(this.bitLength / 8);
    await getRandomBytes(buf);
    return buf;
  }
}

export class IdEncoder {
  /**
   * Creates a new IdEncoder instance.
   *
   * An IdEncoder encodes an array of id bytes into a specific encoding.
   *
   * @param {object} [options] - The options to use.
   * @param {string} [options.encoding='base58'] - Encoding format.
   * @param {boolean} [options.fixedLength=false] - `true` to ensure fixed
   *   output length.
   * @param {number} [options.fixedBitLength] - Fixed output bit length or 0 to
   *   base on input byte size.
   * @param {boolean} [options.multibase=true] - Use multibase encoding.
   * @param {boolean} [options.multihash=false] - Use multihash encoding.
   *
   * @returns {IdEncoder} - New IdEncoder.
   */
  constructor({
    encoding = 'base58',
    fixedLength = false,
    fixedBitLength,
    multibase = true,
    multihash = false,
  } = {}) {
    switch(encoding) {
      case 'hex':
      case 'base16':
        this.encoder = _base16Encoder;
        this.multibasePrefix = 'f';
        break;
      case 'base16upper':
        this.encoder = _base16Encoder;
        this.multibasePrefix = 'F';
        break;
      case 'base58':
      case 'base58btc':
        this.encoder = _base58Encoder;
        this.multibasePrefix = 'z';
        break;
      default:
        throw new Error(`Unknown encoding type: "${encoding}".`);
    }
    this.fixedLength = fixedLength || fixedBitLength !== undefined;
    if(this.fixedLength) {
      this.fixedBitLength = _calcOptionsBitLength({
        // default of 0 calculates from input size
        defaultLength: 0,
        bitLength: fixedBitLength
      });
    }
    this.encoding = encoding;
    this.multibase = multibase;
    this.multihash = multihash;
  }

  /**
   * Encode id bytes into a string.
   *
   * @param {Uint8Array} bytes - Bytes to encode.
   *
   * @returns {string} - Encoded string.
   */
  encode(bytes) {
    if(this.multihash) {
      const byteSize = bytes.length;

      if(byteSize > 127) {
        throw new RangeError('Identifier size too large.');
      }
      // <varint hash fn code> <varint digest size in bytes> <hash fn output>
      //  <identity function>             <byte size>                <raw bytes>
      const multihash = new Uint8Array(2 + byteSize);
      // <varint hash fn code>: identity function
      multihash.set([MULTIHASH_IDENTITY_FUNCTION_CODE]);
      // <varint digest size in bytes>
      multihash.set([byteSize], 1);
      // <hash fn output>: identifier bytes
      multihash.set(bytes, 2);
      bytes = multihash;
    }
    const encoded = this.encoder({bytes, idEncoder: this});
    if(this.multibase) {
      return this.multibasePrefix + encoded;
    }
    return encoded;
  }
}

export class IdDecoder {
  /**
   * Creates a new IdDecoder instance.
   *
   * An IdDecoder decodes an id string into a byte array. It is recommended to
   * use the fixedBitLength option to avoid padding ids resulting in a larger
   * than expected byte length.
   *
   * @param {object} [options] - The options to use.
   * @param {string} [options.encoding='base58'] - Encoding format. Ignored if
   *   multibase is true.
   * @param {number} [options.fixedBitLength] - Fixed output bit length. Values
   *   with leading non-zero data will error.
   * @param {boolean} [options.multibase=true] - Use multibase encoding to
   *   detect the id format.
   * @param {boolean} [options.multihash=false] - Use multihash encoding to
   *   detect the id format.
   * @param {number} [options.expectedSize=32] - Optional expected identifier
   *   size in bytes (only for multihash encoding). Use `0` to disable size
   *   check.
   * @returns {IdDecoder} - New IdDecoder.
   */
  constructor({
    encoding = 'base58',
    fixedBitLength,
    multibase = true,
    multihash = false,
    expectedSize = 32
  } = {}) {
    this.encoding = encoding;
    this.fixedBitLength = fixedBitLength;
    this.multibase = multibase;
    this.multihash = multihash;
    this.expectedSize = expectedSize;
  }

  /**
   * Decode id string into bytes.
   *
   * @param {string} id - Id to decode.
   *
   * @returns {Uint8Array} - Array of decoded id bytes.
   */
  decode(id) {
    let encoding;
    let data;
    if(this.multibase) {
      if(id.length < 1) {
        throw new Error('Multibase encoding not found.');
      }
      const prefix = id[0];
      data = id.substring(1);
      switch(id[0]) {
        case 'f':
          encoding = 'base16';
          break;
        case 'F':
          encoding = 'base16upper';
          break;
        case 'z':
          encoding = 'base58';
          break;
        default:
          throw new Error(`Unknown multibase prefix "${prefix}".`);
      }
    } else {
      encoding = this.encoding;
      data = id;
    }
    let decoded;
    switch(encoding) {
      case 'hex':
      case 'base16':
      case 'base16upper':
        if(data.length % 2 !== 0) {
          throw new Error('Invalid base16 data length.');
        }
        decoded = bytesFromHex(data);
        break;
      case 'base58':
        decoded = base58decoder(data);
        break;
      default:
        throw new Error(`Unknown encoding "${encoding}".`);
    }
    if(!decoded) {
      throw new Error(`Invalid encoded data "${data}".`);
    }
    if(this.fixedBitLength) {
      return _bytesWithBitLength({
        bytes: decoded,
        bitLength: this.fixedBitLength
      });
    }
    if(this.multihash) {
      // <varint hash fn code>: identity function
      const [hashFnCode] = decoded;

      if(hashFnCode !== MULTIHASH_IDENTITY_FUNCTION_CODE) {
        throw new Error('Invalid multihash function code.');
      }
      // <varint digest size in bytes>
      const digestSize = decoded[1];

      if(digestSize > 127) {
        throw new RangeError('Decoded identifier size too large.');
      }

      const bytes = decoded.subarray(2);

      if(bytes.byteLength !== digestSize) {
        throw new RangeError('Unexpected identifier size.');
      }
      if(this.expectedSize && bytes.byteLength !== this.expectedSize) {
        throw new RangeError(
          `Invalid decoded identifier size. Identifier must be ` +
            `"${this.expectedSize}" bytes.`);
      }

      decoded = bytes;
    }
    return decoded;
  }
}

/**
 * Generates an encoded id string from random bits.
 *
 * @param {object} [options] - The options to use. See `IdEncoder` and
 *   `IdGenerator` for available options.
 *
 * @returns {string} - Encoded string id.
 */
export async function generateId(options) {
  return new IdEncoder(options)
    .encode(await new IdGenerator(options).generate());
}

/**
 * Decodes an encoded id string to an array of bytes.
 *
 * @param {object} options - The options to use. See `IdDecoder` for available
 *   options.
 * @param {string} options.id - Id to decode.
 *
 * @returns {Uint8Array} - Decoded array of id bytes.
 */
export function decodeId(options) {
  return new IdDecoder(options).decode(options.id);
}

/**
 * Minimum number of bytes needed to encode an id of a given bit length.
 *
 * @param {object} options - The options to use.
 * @param {string} [options.encoding='base58'] - Encoding format.
 * @param {number} [options.bitLength=128] - Number of id bits.
 * @param {boolean} [options.multibase=true] - Account for multibase encoding.
 *
 * @returns {number} - The minimum number of encoded bytes.
 */
export function minEncodedIdBytes({
  encoding = 'base58',
  bitLength = 128,
  multibase = true
} = {}) {
  let plainBytes;
  switch(encoding) {
    case 'hex':
    case 'base16':
    case 'base16upper':
      plainBytes = bitLength / 4;
      break;
    case 'base58':
    case 'base58btc':
      plainBytes = bitLength / 8;
      break;
    default:
      throw new Error(`Unknown encoding type: "${encoding}".`);
  }
  return plainBytes + (multibase ? 1 : 0);
}

/**
 * Maximum number of bytes needed to encode an id of a given bit length.
 *
 * @param {object} options - The options to use.
 * @param {string} [options.encoding='base58'] - Encoding format.
 * @param {number} [options.bitLength=128] - Number of id bits.
 * @param {boolean} [options.multibase=true] - Account for multibase encoding.
 *
 * @returns {number} - The maximum number of encoded bytes.
 */
export function maxEncodedIdBytes({
  encoding = 'base58',
  bitLength = 128,
  multibase = true
} = {}) {
  let plainBytes;
  switch(encoding) {
    case 'hex':
    case 'base16':
    case 'base16upper':
      plainBytes = bitLength / 4;
      break;
    case 'base58':
    case 'base58btc':
      plainBytes = Math.ceil(bitLength / Math.log2(58));
      break;
    default:
      throw new Error(`Unknown encoding type: "${encoding}".`);
  }
  return plainBytes + (multibase ? 1 : 0);
}

/**
 * Generates a secret key seed encoded as a string that can be stored and later
 * used to generate a key pair. The public key from the key pair can be used as
 * an identifier. The key seed (both raw and encoded form) MUST be kept secret.
 *
 * @param {object} [options] - The options to use.
 * @param {string} [options.encoding='base58'] - Encoding format.
 * @param {number} [options.bitLength=32 * 8] - Number of bits to generate.
 * @param {boolean} [options.multibase=true] - Use multibase encoding.
 * @param {boolean} [options.multihash=true] - Use multihash encoding.

 * @returns {string} - Secret key seed encoded as a string.
 */
export async function generateSecretKeySeed({
  bitLength = 32 * 8,
  encoding = 'base58',
  multibase = true,
  multihash = true
} = {}) {
  // reuse `generateId` for convenience, but a key seed is *SECRET* and
  // not an identifier itself, rather it is used to generate an identifier via
  // a public key
  // Note: Setting fixedLength to false even though that's the (current)
  // default as not using a fixed length of false for a seed is a security
  // problem
  return generateId(
    {bitLength, encoding, fixedLength: false, multibase, multihash});
}

/**
 * Decodes an encoded secret key seed into an array of secret key seed bytes.
 * The key seed bytes MUST be kept secret.
 *
 * @param {object} options - The options to use.
 * @param {boolean} [options.multibase=true] - Use multibase encoding to detect
 *   the id format.
 * @param {boolean} [options.multihash=true] - Use multihash encoding to detect
 *   the id format.
 * @param {number} [options.expectedSize] - Optional expected identifier size
 *   in bytes (only for multihash encoding). Use `0` to disable size check.
 * @param {string} options.secretKeySeed - The secret key seed to be decoded.
 *
 * @returns {Uint8Array} - An array of secret key seed bytes (default size:
 *   32 bytes).
 */
export function decodeSecretKeySeed({
  multibase = true,
  multihash = true,
  expectedSize = 32,
  secretKeySeed,
}) {
  // reuse `decodeId` for convenience, but key seed bytes are *SECRET* and
  // are NOT identifiers, they are used to generate identifiers from public keys
  return decodeId({multihash, multibase, expectedSize, id: secretKeySeed});
}
